From 47e56d65f37e252ea75ec1e176431c0e4998c73c Mon Sep 17 00:00:00 2001
From: Alejandro Abdelnur <tucu@apache.org>
Date: Mon, 11 Mar 2013 18:57:46 +0000
Subject: [PATCH 1292/1357] MR1: MAPREDUCE-5047. keep.failed.task.files=true causes job failure on secure clusters. (sandyr via tucu)

git-svn-id: https://svn.apache.org/repos/asf/hadoop/common/branches/branch-1@1455283 13f79535-47bb-0310-9956-ffa450edef68
(cherry picked from commit b7824899b559389d140415f76a7a830e1a641de7)

Reason: Customer issue
Ref: CDH-4759
Author: Sandy Ryza
---
 .../src/mapred/org/apache/hadoop/mapred/Child.java |    5 +++++
 .../mapred/org/apache/hadoop/mapred/MapTask.java   |    5 ++---
 .../src/mapred/org/apache/hadoop/mapred/Task.java  |    7 +++++++
 3 files changed, 14 insertions(+), 3 deletions(-)

diff --git a/hadoop-mapreduce1-project/src/mapred/org/apache/hadoop/mapred/Child.java b/hadoop-mapreduce1-project/src/mapred/org/apache/hadoop/mapred/Child.java
index 062c7b4..d282861 100644
--- a/hadoop-mapreduce1-project/src/mapred/org/apache/hadoop/mapred/Child.java
+++ b/hadoop-mapreduce1-project/src/mapred/org/apache/hadoop/mapred/Child.java
@@ -339,6 +339,11 @@ class Child {
     // Do the task-type specific localization
     task.localizeConfiguration(jobConf);
     
+    // Write files required to re-run the task with IsolationRunner
+    if (task.supportIsolationRunner(jobConf)) {
+      task.writeFilesRequiredForRerun(jobConf);
+    }
+    
     //write the localized task jobconf
     LocalDirAllocator lDirAlloc = 
       new LocalDirAllocator(JobConf.MAPRED_LOCAL_DIR_PROPERTY);
diff --git a/hadoop-mapreduce1-project/src/mapred/org/apache/hadoop/mapred/MapTask.java b/hadoop-mapreduce1-project/src/mapred/org/apache/hadoop/mapred/MapTask.java
index 99d6d78..49aba02 100644
--- a/hadoop-mapreduce1-project/src/mapred/org/apache/hadoop/mapred/MapTask.java
+++ b/hadoop-mapreduce1-project/src/mapred/org/apache/hadoop/mapred/MapTask.java
@@ -114,15 +114,14 @@ public class MapTask extends Task {
   }
 
   @Override
-  public void localizeConfiguration(JobConf conf)
+  public void writeFilesRequiredForRerun(JobConf conf)
       throws IOException {
-    super.localizeConfiguration(conf);
     // split.info file is used only by IsolationRunner.
     // Write the split file to the local disk if it is a normal map task (not a
     // job-setup or a job-cleanup task) and if the user wishes to run
     // IsolationRunner either by setting keep.failed.tasks.files to true or by
     // using keep.tasks.files.pattern
-    if (supportIsolationRunner(conf) && isMapOrReduce()) {
+    if (isMapOrReduce()) {
       // localize the split meta-information
       Path localSplitMeta =
         new LocalDirAllocator("mapred.local.dir").getLocalPathForWrite(
diff --git a/hadoop-mapreduce1-project/src/mapred/org/apache/hadoop/mapred/Task.java b/hadoop-mapreduce1-project/src/mapred/org/apache/hadoop/mapred/Task.java
index 921f7d2..a28b816 100644
--- a/hadoop-mapreduce1-project/src/mapred/org/apache/hadoop/mapred/Task.java
+++ b/hadoop-mapreduce1-project/src/mapred/org/apache/hadoop/mapred/Task.java
@@ -469,6 +469,13 @@ abstract public class Task implements Writable, Configurable {
     conf.set("mapred.job.id", taskId.getJobID().toString());
   }
   
+  /**
+   * Write files that the IsolationRunner will need to rerun the task.
+   */
+  public void writeFilesRequiredForRerun(JobConf conf) throws IOException {
+    // Do nothing in the general case
+  }
+  
   /** Run this task as a part of the named job.  This method is executed in the
    * child process and is what invokes user-supplied map, reduce, etc. methods.
    * @param umbilical for progress reports
-- 
1.7.0.4

